File: server/data/problems/121-best-time-to-buy-and-sell-stock/solutions/1-precompute-minimums/code.java
class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        if (n == 0) {
            return 0;
        }
        
        // min_prices[i] will hold the minimum price from day 0 to day i
        int[] minPrices = new int[n];
        minPrices[0] = prices[0];
        
        for (int i = 1; i < n; i++) {
            minPrices[i] = Math.min(minPrices[i - 1], prices[i]);
        }
        
        // calculate the profit for selling on each day and keep track of the maximum profit
        int maxProfit = 0;
        for (int i = 0; i < n; i++) {
            int currentProfit = prices[i] - minPrices[i];
            if (currentProfit > maxProfit) {
                maxProfit = currentProfit;
            }
        }
        
        return maxProfit;
    }
}

File: server/data/problems/121-best-time-to-buy-and-sell-stock/solutions/1-precompute-minimums/code.py
from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        if n == 0:
            return 0
        
        min_prices = [0] * n
        min_prices[0] = prices[0]
                
        # min_prices[i] will hold the minimum price from day 0 to day i
        for i in range(1, n):
            if min_prices[i-1] < prices[i]:
                min_prices[i] = min_prices[i-1]
            else:
                min_prices[i] = prices[i]
        
        # calculate the profit for selling on each day and keep track of the maximum profit
        max_profit = 0
        for i in range(n):
            profit = prices[i] - min_prices[i]
            if profit > max_profit:
                max_profit = profit
                
        return max_profit